include/sdf/ParserConfig.hh

```diff
@@ -0,0 +1,130 @@
+#include "sdf/system_util.hh"
+
+namespace sdf
+{
+inline namespace SDF_VERSION_NAMESPACE
+{
+// Forward declare private data class.
+class ParserConfigPrivate;
+
+/// This class contains configuration options for the libsdformat parser.
+///
+/// The configuration options include:
+///   * A callback function used to locate a file from a given URI
+///   * A map from URI schemes to search directories
+///
+/// For backward compatibility, the functions sdf::setFindCallback() and
+/// sdf::addURIPath() update a singleton ParserConfig object, which can be
+/// retrieved by ParserConfig::GlobalConfig().
+///
+/// The functions sdf::readFile(), sdf::readString(), and \ref Root::Load
+/// "sdf::Root::Load()" have overloads that take a ParserConfig object. If
+/// the ParserConfig object is omitted, these functions will use the singleton
+/// ParserConfig object.
+///
+/// Example:
+/// To set an additional URI scheme search directory without affecting the
+/// global config,
+///
+/// \code{.cpp}
+///   // Copy the global config
+///   sdf::ParserConfig config = ParserConfig::GlobalConfig();
+///
+///   // Add the new scheme to the config
+///   config.AddURIPath("newScheme://", "path/to/directory");
+///
+///   // Use the new config when loading a new SDFormat file
+///   sdf::Root root;
+///   root.Load("path/to/file.sdf", config);
+/// \endcode
+class SDFORMAT_VISIBLE ParserConfig
+{
+  /// type alias for the map from URI scheme to search directories
+  public: using SchemeToPathMap =
+          std::map<std::string, std::vector<std::string> >;
+
+  /// \brief Default constructor
+  public: ParserConfig();
+
+  /// \brief Copy constructor
+  /// \param[in] _config ParserConfig to copy.
+  public: ParserConfig(const ParserConfig &_config);
+
+  /// \brief Move constructor
+  /// \param[in] _config ParserConfig to move.
+  public: ParserConfig(ParserConfig &&_config) noexcept;
+
+  /// \brief Move assignment operator.
+  /// \param[in] _config ParserConfig to move.
+  /// \return Reference to this.
+  public: ParserConfig &operator=(ParserConfig &&_config) noexcept;
+
+  /// \brief Copy assignment operator.
+  /// \param[in] _config ParserConfig to copy.
+  /// \return Reference to this.
+  public: ParserConfig &operator=(const ParserConfig &_config);
+
+  /// \brief Destructor
+  public: ~ParserConfig();
+
+  /// Mutable access to a singleton ParserConfig that serves as the global
+  /// ParserConfig object for all parsing operations
+  /// \return A mutable reference to the singleton ParserConfig object
+  public: static ParserConfig &GlobalConfig();
+
+  /// \brief Get the find file callback function
+  /// \return Immutable reference to the find file callback function
+  public: const std::function<std::string(const std::string &)> &
+          FindFileCallback() const;
+
+  /// \brief Set the callback to use when libsdformat can't find a file.
+  /// The callback should return a complete path to the requested file, or
```

Comment:
can you be more explicit about what the input argument to the callback function should be? I believe it can be a file name with no path qualifiers at all, but can it be a relative path? I doubt it would ever be an absolute path based on "when libsdformat can't find a file", but it would be helpful to be explicit about what are the expected inputs for that callback function

Reply:
I don't think there's any limitation as to what the input argument would be. It really depends on the users callback function. `sdf::findFile` calls the function if all other ways of searching for the file fail. This means it can even be called with absolute paths if the file pointed to by the path doesn't exist on the filesystem.

Reply:
I think that's still useful to explain, that it's up to the callback to interpret the input arguments. Please correct me if I'm wrong, but perhaps we could give an example that if a file has text like `custom://model_name` in a `//include/uri` tag, that text should eventually be passed to this callback function if a file with that name can't be found?

Reply:
I added an explanation in https://github.com/osrf/sdformat/pull/439/commits/8a9e4e82e99d5a3760b02a5be56dbae2a02f3008

Reply:
:+1: